Naveego.Streaming.Kafka/KafkaStreamReader.cs

```diff
@@ -55,68 +51,98 @@ private async void Run(Func<T, Task<HandleResult>> onMessage, CancellationToken
-                                    break;
-                                }
-                            }
-                        }
-                        catch (ConsumeException e)
-                        {
-                            Logger.LogError(e, $"Error processing kafka stream: {e.Message}");
-                        }
+                        await ConsumeAndProcessMessage(onMessage, c, cancellationToken);
                     }
 
                     // Ensure the consumer leaves the group cleanly and final offsets are committed.
                     c.Close();
                 }
             }
+            
             catch (Exception ex)
             {
                 Logger.LogError(ex, $"Could not create consumer on Kafka stream: {ex.Message}");
                 throw;
             }
         }
+
+        private async Task ConsumeAndProcessMessage(
+            Func<T, Task<HandleResult>> onMessage,
+            Consumer<Ignore, string> c,
+            CancellationToken cancellationToken)
+        {
+            // Check the cancellation token to see if we need to stop
+                cancellationToken.ThrowIfCancellationRequested();
+                try
+                {
+                    var cr = c.Consume();
+                    T item;
+                    using (var jsonReader = new JsonTextReader(new StringReader(cr.Value)))
+                    {
+                        item = _serializer.Deserialize<T>(jsonReader);
+                    }
+
+                    var result = await onMessage(item);
+                    
+                    // If the result was a success then commit the offsets
+                    // TODO: Improve this to commit offsets in background for performance reasons
+                    if (result.Success)
+                    {
+                        c.Commit();
+                        // move on
+                        return;
+                    }
+                    
+                    Logger.LogWarning("Processing of message was not successful.  Retrying...");
+
+                    await RetryProcessingMessage(item, onMessage, c, cancellationToken, result);
+                }
+                catch (Exception e)
+                {
+                    switch (e)
+                    {
+                        case ConsumeException exception:
+                            Logger.LogError(e, $"Consuming kafka message error: {e.Message}");
+                            break;
+                        case TopicPartitionOffsetException exception:
+                            Logger.LogError(e, $"Committing topic offset error: {e.Message}");
+                            break;
+                        case KafkaException exception:
+                            Logger.LogError(e, $"Error interacting with kafka: {e.Message}");
+                            break;
+                        default:
+                            throw;
```

Comment:
What happens when this propagates the exception?  It looks like it will break the while loop in the `ReadAsync` enclosing method.  Does this cause a problem with the caller?  It is up to them to recover and restart the consumer?

Reply:
How does the Kafka library handle the situation where the consumer is re-created?  Is that where partition assignment occurs?  Wouldn't we be exiting and re-entering the `using (var c = new Consumer<Ignore, string>(_config))` block?  Just wondering if the caller might get re-connected to an entirely different partition, or if that even matters.

Reply:
When a consumer enters or leaves the consumer group, Kafka triggers a consumer re-balance. So it doesn't matter, since when consumers leave or enter, Kafka re-balances the consumers per partition each time.

Your talking about the `default: thow;` case right? I guess my point is that if it gets to that point the consumer should blow up because it's the `onMessage`'s exception.